chain/store/store.go

```diff
@@ -730,38 +737,82 @@ func (cs *ChainStore) MessagesForTipset(ts *types.TipSet) ([]types.ChainMsg, err
 		return nil
 	}
 
-	var out []types.ChainMsg
+	selectMsg := func(m *types.Message) (bool, error) {
+		if err := preloadAddr(m.From); err != nil {
+			return false, err
+		}
+
+		if applied[m.From] != m.Nonce {
+			return false, nil
```

Comment:
Messages with bad nonce should also be applied so miner pays the penalty.

Reply:
They can't end up in blocks: https://github.com/filecoin-project/lotus/blob/bf774fc29b6910cadc4a21726aa68d324a54df7a/chain/sync.go#L1010-L1028